Better environment information

diegoreymendez opened the pull request:
### Description:

Adds a few improvements to our mechanisms for obtaining information about the build and device the code is running in.

### Details:

Improvements:

- Adds [testing override capabilities](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896/files#diff-9fd2cb6f296e8b3a8116115a801f90c5R29) to our methods that check if the code is running in the simulator.
- Makes for [more readable](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896/files#diff-7886af00b902aa11caa91e567c363188R46) device-type and build-configuration checks.
- [Separates the old Build enum](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896/files#diff-b4a43aa0802b2b958f507210c2e1cbb8L39) from the `FeatureFlag` file.  While related, those were two separate entities.
- Adds Obj-C support for [device type overriding](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896/files#diff-972bf7c19ca93aa01d08fd6e06c5e6f2R35) during testing.
- Removes [a method that wasn't really testing anything](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896/files#diff-48f25c7a1fb48d0f8d2211a81dcea36eL26) and was only increasing the number of unit tests we have.  If needed it can be added back as soon as we have tests for it.

### Testing:

**Test 1:**

We have new code that conditionally compiles depending on the target.
Build WPiOS in debug, release-internal, release-alpha and release configurations.

Note: unit tests are not really prepared to run in anything other than debug configuration, just make sure WPiOS builds in those configs.

**Test 2:**

Check that the `pluginManagement` feature flag is working for `localDeveloper` only (@koke)
Check that the `jetpackDisconnect` feature flag is working for `localDeveloper` only (@elibud)
Check that the `googleLogin` feature flag is working for `localDeveloper` only (@nheagy)

diegoreymendez commented:
Adding 3 reviewers just to make sure the changes are fine.

diegoreymendez commented on WordPress/WordPressTest/EditorSettingsTests.swift:
```diff
@@ -11,17 +11,19 @@ class EditorSettingsTests: XCTestCase {
+            editorSettings.nativeEditorEnabled = true
 
-        XCTAssertTrue(editorSettings.nativeEditorEnabled)
+            XCTAssertTrue(editorSettings.nativeEditorEnabled)
```
Would be great information if the line had trailing whitespace.

koke commented:
> Makes for more readable device-type and build-configuration checks

I'm not loving the new `Build.is` syntax. It's readable in the sense that it reads more like English but somehow doesn't seem to fit the Swift style model. `Build` is a type, not a thing and so it can't `be` something. I agree that the existing calls are also not ideal, and maybe more like C than Swift.

So I have another suggestion:

```swift
// # Single configuration test
// Before
Build.is(.localDeveloper) 
// After
build(is: .localDeveloper) 

// # Multiple configuration test
// Before
Build.is([.localDeveloper, .a8cBranchTest]) 
// After 
build(anyOf: [.localDeveloper, .a8cBranchTest]) 
```

koke commented on WordPress/Classes/Utility/BuildInformation/Device.swift:
```diff
@@ -0,0 +1,40 @@
+///
+@objc
+class Device: NSObject {
+    @objc
```
IIRC, nested types can't be visible to Obj-C anyway, so no need for the `@objc`

diegoreymendez commented:
So a few thoughts on all this, because I've spent quite a bit of time yesterday thinking about it:

1. Right now we have quite a few methods and properties wrapped inside the `Build` and `Device` static classes.  The current solution doesn't need to be the one we keep, but I'd encourage whatever solution we pick to exist inside it's own class at least (as opposed to a global `build()` method).

2. I think `buildConfiguration` may be better wording than `build` for the check we're performing.

3. Another alternative to at least keep this information inside its own class, but sort of follow the pattern you propose could be to try and group all the classes and configurations inside something like:

```swift
Environment.buildConfiguration(is: .localDeveloper)
Environment.device(is: .simulator)
```

This would have the added advantage that the class could eventually be converted to a singleton, allowing us to move it out of the global scope if we so wanted.

elibud commented:
@diegoreymendez I really like your idea of keeping everything in an `Environment` class. 
I did a quick pass and this lgtm but lets wait to see what everyone else has to say, and if we decide to change the naming (I'm also not crazy about `Build.is`) I can re-test this thoroughly.

diegoreymendez commented:
I think the observation by @koke that `Build` is not an object but a class is a reasonable one one.  It warrants a change in the code.

On a related note, in general static classes and static methods are also undesirable.  I think the Environment static class (or even better singleton) could help us keep this to a minimum, so I'll be changing this PR towards that tomorrow.

I'll wait until tomorrow to move on if there's no more input in this PR in order to see if there's consensus, but also give it a deadline to keep things dynamic.

diegoreymendez commented:
PS: our code checking if it's running in a simulator or not is a code smell... and I think we should eventually remove those checks.

nheagy commented on WordPress/Classes/Utility/BuildInformation/Device.swift:
```diff
@@ -0,0 +1,40 @@
+/// Runtime information about the current device.
+///
+@objc
+class Device: NSObject {
```
I'd expect an object called `Device` to know about iPad vs iPhone vs Watch, and be surprised to find it about simulator vs physical. Any thoughts on renaming this to something like `Runtime` or `DeviceRuntime`?

diegoreymendez replied:
Considering my task after wrapping this up will be to completely eliminate the need for this, `DeviceRuntime` sounds about right to me. :)

Sounds good?

nheagy replied:
@diegoreymendez sounds great imo!

nheagy commented on WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift:
```diff
@@ -0,0 +1,36 @@
+/// Objective-C bridge for FeatureFlag.
+///
+/// Since we can't expose properties on Swift enums we use a class instead
+class Feature: NSObject {
```
I don't see any uses of this in ObjC. Do we have to expose this to ObjC?

koke replied:
I see at least three:

![screen shot 2017-10-04 at 09 19 21](https://user-images.githubusercontent.com/8739/31163946-34873f38-a8e5-11e7-96e1-39fe400f1728.png)

nheagy replied:
Well damn, I completely botched my search.

nheagy commented on WordPress/Classes/Utility/BuildInformation/Build.swift:
```diff
@@ -0,0 +1,73 @@
+@objc
+class Build: NSObject {
```
Is this used in ObjC anywhere?

diegoreymendez replied:
It's not, although I ended up putting a lot of effort into making the code Obj-C compatible since we have so much Obj-C code in our app, and our unit tests could end up needing it.

This is one of the reasons why I think we should make it a priority to migrate our code to Swift.

At the same time I'd hate this code being duplicated just because it didn't work for Obj-C (if that makes any sense).

koke replied:
The original reason (maybe not a great one) for keeping `Build` inside the `FeatureFlag.swift` was that it was meant as a helper for the feature flags. As you mentioned somewhere else, the other types can be considered _code smells_, and I'd rather not have anything other than `Feature` check for the build configuration.

diegoreymendez replied:
That sounds about right, but that will affect how we test that class.  Overriding the value of a private detail should not be done by automated tests.

koke replied:
Wouldn't [this](https://github.com/wordpress-mobile/WordPress-iOS/pull/7896#issuecomment-334075022) be enough for testing?

diegoreymendez replied:
I'm not saying that's not enough... but [these](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/WordPressTest/FeatureFlagTest.swift#L8) tests involve manipulating `Build` information.

The approach used in those tests to override the `Build` information is one that I particularly don't like.  In fact for cases like that I find that dependency injection is a much more appropriate approach.  But dependency injection would require `Build` not to be a private class.

nheagy commented on WordPress/Classes/Utility/BuildInformation/Build.swift:
```diff
@@ -0,0 +1,73 @@
+    ///
+    ///     let enableExperimentalStuff = build(.localDeveloper, .a8cBranchTest)
+    ///
+    static func `is`(_ testConfiguration: BuildConfiguration) -> Bool {
```
The backticks to name this `is` make me ponder, but I think fundamentally if this were a Swift enum instead of a class—if we don't need to use this in ObjC, as I suspect we may not—we could use built-in equality to evaluate this instead of providing a method?

diegoreymendez replied:
Kind of related to what I said above.  My take on this is that as long as we have Obj-C code and tests, making these methods and classes Swift-only would be not ideal.

nheagy replied:
I understand the idea that _maybe_ we'll do new development with Objective-C, but atm our principle is that all new code is Swift, making the chance that we have to support ObjC pretty slim. fwiw, I vote we do this the nice Swifty way 😜

koke commented:
Quoting @nheagy:

> if this were a Swift enum instead of a class—if we don't need to use this in ObjC, as I suspect we may not—we could use built-in equality to evaluate this instead of providing a method?

I think that might work even better than an `is` or a function. Make `BuildConfiguration` the enum, and make it have a `current` property. You can even use the standard pattern matching operator `~=` to test against multiple build configurations:


```swift
enum BuildConfiguration {
    /// Development build, usually run from Xcode
    case localDeveloper
    /// Continuous integration builds for Automattic employees to test branches & PRs
    case a8cBranchTest
    /// Beta released internally for Automattic employees
    case a8cPrereleaseTesting
    /// Production build released in the app store
    case appStore

    static var current: BuildConfiguration {
        // Mock result
        return .localDeveloper
    }

    static func ~=(a: BuildConfiguration, b: Set<BuildConfiguration>) -> Bool {
        return b.contains(a)
    }
}

BuildConfiguration.current == .localDeveloper
BuildConfiguration.current ~= [.localDeveloper, .a8cBranchTest]
```

koke commented:
Another suggestion to limit all the testing overrides. I think with the proposed approach, `BuildTests` are kind of superfluous, since you'd be basically testing the equality operator 😬

For `FeatureFlag` testing, we could do something like this:

```swift
enum FeatureFlag: Int {
    case exampleFeature
    case newMediaExports
    case pluginManagement
    case googleLogin
    case jetpackDisconnect

    func enabled(in buildConfiguration: BuildConfiguration = BuildConfiguration.current) -> Bool {
        switch self {
        case .exampleFeature:
            return true
        case .newMediaExports:
            return buildConfiguration ~= [.localDeveloper, .a8cBranchTest]
        case .pluginManagement:
            return buildConfiguration == .localDeveloper
        case .googleLogin:
            return buildConfiguration == .localDeveloper
        case .jetpackDisconnect:
            return buildConfiguration == .localDeveloper
        }
    }
}
```

diegoreymendez commented:
@koke - Going all Swift would be great, but these utilities are generic enough that we may still need to access them from our Obj-C code.  If we make them inaccessible from Obj-C and someone needs to use them, we'll end up with duplicated code to achieve the same thing.

The solution proposed in this PR can be improved and / or changed, but any drastically different suggestions should be tested in a PR in order to make sure the solution does work for both languages.

PS: this is not an attempt to kill the conversation, feel free to link to other PRs here and keep the conversation going.

elibud commented:
@koke 
> You can even use the standard pattern matching operator `~=`

I had no idea, that's so brilliant!

diegoreymendez commented:
Ping @nheagy, @koke: So as I understand it the idea is that since we're not doing new developments in Obj-C we'll never ever have to check against `BuildConfiguration` from Obj-C.  That's a point that makes sense and I missed before.

I'll modify this PR to propose a Swift solution.

jleandroperez commented:
+1000. New ObjC should be strongly discouraged (if not, forbidden!). Except, right, minor maintenance over existing code.

diegoreymendez commented:
@nheagy - Just to clarify, I haven't uploaded any changes because I'm currently working on several different things, but I'll make some changes in the next few days.

Thanks everyone for the awesome feedback!

diegoreymendez commented:
@nheagy, @koke - Ready for another round.

Changes:

1. Our code no longer checks if it's running in a simulator or device.
2. Thanks to point 1, we no longer need Device.swift.
3. Renamed `Build` to `BuildConfiguration` and used an approach similar to what @koke proposed.
4. Added an improved mechanism for testing with build configurations.

Current limitations we may want to evaluate:

a. We cannot make `BuildConfiguration` private because we have unit tests using it.
b. We cannot make `FeatureFlag` swift-only because there's Obj-C making use of it.
